### PR TITLE
Initial skeleton of some component model processing

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is the first of what will likely be many to implement the
component model proposal in Wasmtime. This will be structured as a
series of incremental commits, most of which haven't been written yet.
My hope is to make this incremental and over time to make this easier to
review and easier to test each step in isolation.

Here much of the skeleton of how components are going to work in
Wasmtime is sketched out. This is not a complete implementation of the
component model so it's not all that useful yet, but some things you can
do are:

* Process the type section into a representation amenable for working
  with in Wasmtime.
* Process the module section and register core wasm modules.
* Process the instance section for core wasm modules.
* Process core wasm module imports.
* Process core wasm instance aliasing.
* Ability to compile a component with core wasm embedded.
* Ability to instantiate a component with no imports.
* Ability to get functions from this component.

This is already starting to diverge from the previous module linking
representation where a `Component` will try to avoid unnecessary
metadata about the component and instead internally only have the bare
minimum necessary to instantiate the module. My hope is we can avoid
constructing most of the index spaces during instantiation only for it
to all ge thrown away. Additionally I'm predicting that we'll need to
see through processing where possible to know how to generate adapters
and where they are fused.

At this time you can't actually call a component's functions, and that's
the next PR that I would like to make.

### CODE REVIEW

**Comment by alexcrichton** [comment]
It's worth pointing out that this is starting out as a draft PR. The main reason for this is that there aren't actually any tests in this PR for that. For tests I'm think we'll need https://github.com/bytecodealliance/wasm-tools/pull/529 which is the text format for components. Tests can probably be written in the meantime with `wasm-encoder` but I'm hoping we won't have to resort to that.

File: crates/environ/src/component/types.rs
Comment (fitzgen) [inline crates/environ/src/component/types.rs:446]: Can this be a `From` implementation?

File: crates/wasmtime/src/instance.rs
Comment (fitzgen) [inline crates/wasmtime/src/instance.rs:572]:
```suggestion
    /// Note that this is unsafe as the validity of `item` is not verified and
```

**Review by fitzgen — COMMENTED** [review]
LGTM modulo tests! I'll take another look at this once those are written

File: crates/environ/src/component/translate.rs
Comment (alexcrichton) [reply crates/environ/src/component/translate.rs:295]: Consider me [sufficiently inspired](https://github.com/bytecodealliance/wasm-tools/pull/556)

**Comment by alexcrichton** [comment]
Ok with https://github.com/bytecodealliance/wasm-tools/pull/529 landed I've now started writing tests for this PR. Once all the pending PRs on wasm-tools are landed I'll make a release and then this should be ready-for-landing after another round of review.

**Comment by alexcrichton** [comment]
Ok everything is now ready to go here and I think that this is good to land.

@fitzgen would you be up for reviewing the tests added here? (sorry I know this is quite late relative to the previous review and this is probably pretty far out of cache for you...)

**Comment by fitzgen** [comment]
Yeah I can take a look tomorrow.

**Review by fitzgen — APPROVED** [review]
Looks great!

**Pull request merged**
